pkg/datasource/data.go

```diff
@@ -137,9 +136,25 @@ func (ds *dataSource) NewData() Data {
 	d := &data{
 		Payload: make([][]byte, ds.payloadCount),
 	}
-	for i := range d.Payload {
-		d.Payload[i] = make([]byte, 0)
+
+	// Allocate memory for fixed size fields added with Add{Sub}Field
```

Comment:
Maybe it's time to re-add the `sync.Pool` and then just reset the contents on the individual fields after getting an object from the pool. Can be done in another PR, though.

Also, we should at least check the sizes when receiving data from gRPC to give the same write guarantees on the client - or drop the packet (can also be done on a later PR).

Reply:
It turns out that this PR broke the gadgectl / kubectl-gadget cases, an invalid memory access panic happens on `NewData()` because `PayloadIndex` is not set on the client, (it's not carried on the protobuf message), then the `d.Payload[f.PayloadIndex] = make([]byte, foo)` line crashes. 

I propose a solution to this in the last commit I pushed. WDYT?

Reply:
> because PayloadIndex is not set on the client, (it's not carried on the protobuf message)

Hmm, I fail to see how/why it's not there as it's set and also part of the protobuf message. It should always be present unless `FieldFlagEmpty` is set, in which case the field doesn't have its own payload. I think it's important to fix it if that's not the case. I'll take a closer look later on again.

What I can see missing from the protobuf message is the `payloadCount` - when deserializing the DataSource info, it should be set to the expected payloadCount - as that is also necessary for adding new fields in operators on the client.

(I'm fine with having separate functions, though.)

Reply:
> Hmm, I fail to see how/why it's not there as it's set and also part of the protobuf message. It should always be present unless FieldFlagEmpty is set, in which case the field doesn't have its own payload. I think it's important to fix it if that's not the case. I'll take a closer look later on again.

> What I can see missing from the protobuf message is the payloadCount - when deserializing the DataSource info, it should be set to the expected payloadCount - as that is also necessary for adding new fields in operators on the client.

I wrote this too fast. It's the `payloadCount` member on the datasource that is not set, hence the `Payload: make([][]byte, ds.payloadCount)` produces an empty payload, then the `d.Payload[f.PayloadIndex] = make([]byte, foo)` crashes as `d.Payload[f.PayloadIndex]` is an invalid memory access.

I agree we could carry this field over, but I think we need to use a different function to unmarshal from protobuf, otherwise all those allocations will be useless.

Reply:
> I propose a solution to this in the last commit I pushed. WDYT?
> 
It looks fine to me.

> I think we need to use a different function to unmarshal from protobuf, otherwise all those allocations will be useless.
>
I agree with the dedicated function you created for unmarshal but why do you say that the allocation will be useless if we don't pre-allocate things? While I do see the advantage of pre-allocating things on the emitter (we make the API cleaner than [this](https://github.com/inspektor-gadget/inspektor-gadget/blob/c35c13c7dbfaa0f0b2781cf38d4f5c8a4af17a09/pkg/operators/ebpf/snapshotter.go#L194-L196)), I don't understand what's the advantage of pre-allocating things on the receiver if `proto.Unmarshal` will do that for us. What am I missing?

Regarding the `payloadCount`, I think that sending it from the server gives us a way to ensure we received the expected number of rows in the payload. Don't you think?

Reply:
> Regarding the payloadCount, I think that sending it from the server gives us a way to ensure we received the expected number of rows in the payload. Don't you think?

This wouldn't ensure it (remote end could just lie about it) - we always have to check anyway. The DataSource info from the GadgetInfo is basically the contract that we have to check everything against. The `payloadCount` can be derived from that, so if we have to check for that anyway, we don't need to send it.


> I agree with the dedicated function you created for unmarshal but why do you say that the allocation will be useless if we don't pre-allocate things? While I do see the advantage of pre-allocating things on the emitter (we make the API cleaner than [this](https://github.com/inspektor-gadget/inspektor-gadget/blob/c35c13c7dbfaa0f0b2781cf38d4f5c8a4af17a09/pkg/operators/ebpf/snapshotter.go#L194-L196)), I don't understand what's the advantage of pre-allocating things on the receiver if proto.Unmarshal will do that for us. What am I missing?

You're probably talking about the same: when the event is coming from gRPC, it makes no sense to pre-allocate as gRPC will allocate anyway - hence the two different functions to initialize. This second deserialize function still has to check lengths against the "contract" to allow that data to pass through, otherwise things could crash.